pydicom/pixel_data_handlers/pillow_handler.py

```diff
@@ -89,49 +96,130 @@ def should_change_PhotometricInterpretation_to_RGB(ds: "Dataset") -> bool:
+    #     1: YCbCr (3 components)
+    #     2: YCCK (4 components)
+    # In Image.info, "adobe_transform" is the APP14 AP12 value as int
+    if "adobe_transform" in im.info:
+        cs = "RGB" if im.info["adobe_transform"] == 0 else "YCbCr"
+
+    # We want to return the pixel data in the color space specified by
+    #   *Photometric Interpretation*, with a warning if that color space
+    #   doesn't actually match that of the encoded source data
+
+    # The possibilities are:
+    #   cs    | PI  -> transform    | return | warn
+    #   ------+---------------------+--------+-----
+    #   None  | RGB -> YCbCr to RGB | RGB    |
+    #   YCbCr | RGB -> YCbCr to RGB | RGB    | yes
+    #   RGB   | RGB -> (none)       | RGB    |
+    #   None  | YBR -> (none)       | YBR    |
+    #   YCbCr | YBR -> (none)       | YBR    |
+    #   RGB   | YBR -> RGB to YCbCr | YBR    | yes
+
+    if photometric_interpretation == "RGB":
+        if cs == "RGB":
+            # Source data is RGB - allow libjpeg to skip transform
+            # If "adobe_transform" is present, or if the component IDs are
+            #   RGB then libjpeg/turbo will use that info to skip the
+            #   YCbCr -> RGB transform
+            # https://github.com/libjpeg-turbo/ijg/blob/main/jdapimin.c
+            pass
+
+        if cs is None:
+            # Source data may be YCbCr (most likely) or RGB (less likely)
+            # libjpeg assumes source is YCbCr - force no transform
+            # Assume *Photometric Interpretation* is correct:
+            # * If the decoded pixel data is correct then source is RGB
+            # * If the decoded pixel data is incorrect then source is YCbCr
+            #   but this can be fixed by user applying YCbCr -> RGB transform
+            im.draft("YCbCr", (shape[0], shape[1]))
+
+        if cs == "YCbCr":
+            # Source data is YCbCr - allow libjpeg transform to RGB and warn
+            # YBR_FULL_422 is only a valid recommendation if we don't
+            #   support JPEGLosslessP14 and JPEGLosslessSV1
+            warnings.warn(
+                "A mismatch was found between the JPEG codestream and dataset "
+                "'Photometric Interpretation' value. If the decoded pixel "
+                "data is in RGB color space then the (0028,0004) "
+                "'Photometric Interpretation' should be 'YBR_FULL_422' "
+                "instead of 'RGB'"
+            )
+
+        return im
+
+    if cs == "RGB":
+        # Source data is RGB but PI is YBR
+        # Trying to forcibly decode to YCbCr results in a
+        #   OSError: broken data stream when reading image file
+        raise AttributeError(
+            "Unable to decode as the JPEG codestream indicates the encoded "
+            "pixel data is in RGB color space, but the (0028,0004) "
+            f"'Photometric Interpretation' is '{photometric_interpretation}'. "
+            "Change the 'Photometric Interpretation' to 'RGB'"
+        )
```

Comment:
It might be better to warn here and return the pixel data (as RGB, unfortunately) rather than raising, since the exception message may just get lost if multiple handlers are available.